R/phylo_check.R

```diff
@@ -0,0 +1,151 @@
+#' out = "tree")
+#' plot(my_ceratopsians)
+#' @export
+
+phylo_check <- function(tree = NULL, list = NULL, out = "table",
+                        sort = "presence") {
+  #Errors for incorrect input
+  if (is.null(tree)) {
+    stop("Phylogeny must be provided")
+  }
+
+  if (inherits(tree, "phylo") == FALSE) {
+    stop("Phylogeny must be a phylo object")
+  }
+
+  if (is.null(list)) {
+    stop("List of taxa to check against must be provided")
+  }
+
+  if (is.vector(list) == FALSE) {
+    stop("List of taxa must be a vector")
+  }
+
+  if (any(grepl("[^[:alnum:][:space:]_]", list))) {
+    stop("Taxon names should not contain punctuation except spaces or
+         underscores")
+  }
+
+  if (out != "counts" && out != "table" && out != "tree") {
+    stop("out must either be 'table', 'counts' or 'tree'")
+  }
+
+  if (sort != "az" && sort != "presence") {
+    stop("sort must either be 'az' or 'presence'")
+  }
+
+  if (out != "table" && sort != "presence") {
+    warning("sort is redundant when using outputs other than 'table'")
+  }
+
+  #Function
+  #Replace any spaces in taxon names with underscores
+  list <- gsub(" ", "_", list)
+  tree$tip.label <- gsub(" ", "_", tree$tip.label)
+
+  #Give uniform capitalisation
+  list <- gsub("^([a-z])", "\\U\\1", tolower(list), perl = TRUE)
+  tree$tip.label <- gsub("^([a-z])", "\\U\\1", tolower(tree$tip.label),
+                         perl = TRUE)
+
+  #Create vectors of names, those in tree and those in list
+  if (out == "counts" || out == "table") {
+    tip_names <- tree$tip.label
+    all_names <- unique(c(tip_names, list))
+    names_in_tree <- (all_names %in% tip_names)
+    names_in_list <- (all_names %in% list)
+  }
+
+  #Determine which names are in which lists and table them
+  if (out == "table") {
+    table <- data.frame(all_names, names_in_tree, names_in_list)
+
+    if (sort == "az") {
+      table <- table[order(table$all_names), ]
+    } else {
+      table <- table[order(table$names_in_list, decreasing = TRUE), ]
+    }
+
+    colnames(table) <- c("Taxon name", "Present in tree", "Present in list")
+
+    return(table)
+  }
+
+  #Determine which names are in which lists and sum them up
+  if (out == "counts") {
+    tree_counts <- as.numeric(names_in_tree)
+    list_counts <- as.numeric(names_in_list)
+    difference <- tree_counts - list_counts
+
+    only_tree <- length(which(difference == 1))
+    only_list <- length(which(difference == -1))
+    in_both <- length(which(difference == 0))
+
+    counts <- data.frame(c("Tree and list", "Only in tree", "Only in list"),
+                        c(in_both, only_tree, only_list))
+    colnames(counts) <- c("Category", "Number of taxa")
+
+    return(counts)
+  }
+
+  #Trim names not in the list from the phylogeny
+  if (out == "tree") {
+    list_in_tree <- match(tree$tip.label, list)
+    no_match <- tree$tip.label[which(is.na(list_in_tree))]
+    smaller_tree <- drop.tip(tree, no_match)
```

Comment:
First of all, I'm not sure I'm commenting in the right section. Sorry, it's my first time doing this!
About the `drop.tip` function used here, I would recommend using paleotree's `DropPaleoTip`. This is because `drop.tip` messes up with branch lengths of non-ultrametric trees and `DropPaleoTip` kind of keeps the same node ages (of the remaining nodes). Oh, and it looks like `DropPaleoTip` was recently merged into function `modifyTerminalBranches`, but  apparently `DropPaleoTip` will still work...

Reply:
Note that `DropPaleoTip` requires the phylogeny to have a $root.time attribute, which may not always be the case. In fact, all it does is call `drop.tip` then call `fixRootTime`, so I would say that we could mention in the details that users may want to call `fixRootTime` after running this function if their tree is not ultrametric.

Reply:
Sounds like a great solution! Thanks @willgearty!